Fix user's shipping & billing address interaction with order

mrbrdo opened the pull request:
There are several issues regarding this, see the individual commits.

This also makes legacy frontend work correctly:
- when order is created, the billing/shipping address will correctly be assigned to user if he doesn't have an address assigned
- when a new order is started, it will correctly use the user's billing/shipping address if present

**This was quite broken with even a security issue with the unscoped Spree::Address query. I suggest to merge asap due to this.**

mrbrdo committed:
new order uses shipping and billing address of user - fix so it's actually used

mrbrdo committed:
fix address empty? since even with a user_id it should be considered empty, and preferences can be {} so check for blank?

mrbrdo committed:
fix issue where order can be assigned an address from a different user and also potentially a deleted address

mrbrdo committed:
user's billing/shipping address will not update when order is updated, fix this

It would be more proper to put this into an Order after_save hook, but the same can be said for
calling the update_or_create_address method, so I will leave that to a future refactor

stefnnn commented:
This is tackling simliar problems to my PR here: https://github.com/spree/spree/pull/11606, maybe we want to merge this into one PR?

mrbrdo commented:
@stefnnn yes we could. Although your PR was not merged after 2 months. I am afraid the maintainers are sleeping?
No point in putting in the work if it will never get merged.

mrbrdo commented:
@damianlegawiec ping, this is kinda important

mrbrdo committed:
When creating Order, correctly assign ship_address or use_billing

The Order should use_billing == true if ship_address is missing, is empty, or is the same as bill_address

mrbrdo commented:
Updated to correctly determine use_billing? (fixes issue if ship is same as bill and gets updated in checkout form)

mrbrdo committed:
don't assign ship_address to order if use_billing was checked in checkout

mrbrdo commented:
@damianlegawiec what do you think?

mrbrdo commented:
@damianlegawiec well??
